### ISSUE TITLE
ORA remote `whereis` reports wrong HTTP URL

### ISSUE THREAD

**Issue by mih** [datalad/datalad#6086]
Such a URL is reported

https://datapub.fz-juelich.de/studyforrest/studyforrest.ria/annex/objects/Kz/7V/MD5E-s37132--f8a50b1462dbb3eb403013d28c197a91.nii.gz/MD5E-s37132--f8a50b1462dbb3eb403013d28c197a91.nii.gz

but the actual URL is 

https://datapub.fz-juelich.de/studyforrest/studyforrest.ria/7fc/d8812-d0fe-11e7-8db2-a0369f7c647e/annex/objects/Kz/7V/MD5E-s37132--f8a50b1462dbb3eb403013d28c197a91.nii.gz/MD5E-s37132--f8a50b1462dbb3eb403013d28c197a91.nii.gz

(dataset ID part inserted)

edit by @yarikoptic : fixed the "actual" by adding `/` before `annex`

**Comment by mih** [comment]
This seems to be a reincarnation of #4979.

**Comment by mih** [comment]
Discussed in a call. There was basic agreement that broken URLs should not be reported.

@bpoldrack suggested to included a hardcoded `dsid[:3]/dsid[3:]` in the output, and veto'ed the use of `dsobj_dir`. Likewise, he objected the use of `key_path`.

@mih still doesn't understand why this would be a sensible move. Hardcoding `dsid[:3]/dsid[3:]` would perform layout version independent inference, while `dsobj_dir` is based on the actual store configuration. Using `key_path` also is based on actual store configuration, whereas hardcoding a directory based on `dirhash` is not. It is unclear why `ssh-based` access can work out of the box, while http-based access should required particular server-side URL mapping, beyond mere exposure of a store, in order to make HTTP. To make my confusion complete, the `transfer_retrieve()` implementation does `dsobj_dir / key_path` -- which is conceptually exactly what the veto'ed change does.

NB: This path handling is POSIX-only, and would yield incorrect HTTP URLs on non-Posix systems.

Because these aspects remain unresolved, a maintenance fix is to disable the reporting of incorrect URLs, by not reporting URLs. A correct fix can then be introduced later.

**Comment by bpoldrack** [comment]
@mih 
> @bpoldrack suggested to included a hardcoded dsid[:3]/dsid[3:] in the output, and veto'ed the use of dsobj_dir. Likewise, he objected the use of key_path.

Not exactly. It's about `key_path` only. 


Let me try again: It's only about being conceptually clear what is the thing we want. I don't veto anything - I'm fine with both approaches. They do have different implications, though. All I want is us being clear about what they are and make an (hopefully informed) decision, which then can go into a specification of what are the requests a server serving a RIA store over HTTP is expected to respond to. Not least, because if memory doesn't entirely refuse to serve, it was you a ong time ago, suggesting a fixed request would be better. So, if you changed your mind I wanted to make sure you realize this.

Also, regarding your last mentioned confusion: That was my confusion, looking locally on an older branch for the parts I don't see in the diff of your PR. So, you are right: It already changed at the end of `get` and `whereis` is out of sync. However, I think that change was not fully thought through when introduced and so I stuck to the question regardless, since I simply didn't want to finalize that move by merging your fix without having a clear decision for what is the target.
So, from my POV: Forget what's to what extend implemented right now and buggy or not - "what is it that it should be" is what I want to clarify.

The main thing to realize, I think, is that all the layout stuff was introduced with SSH and local access in mind, both of which are "remote execution" approaches. They actually operate on the file system where this layout is used. HTTP is conceptually fundamentally different. That's why a mere "this function is what ORA thinks" doesn't really apply. "This function is what ORA thought, when there was nothing but remote execution variants" is more accurate.


- Option 1:
Initially (as per PR #4561) we went for a fixed request URL, that does not consider the store-side layout. Instead, we made it the server's responsibility to deal with that. That's kind of a "service" point of view. We have a fixed request from a) a store, b) a dataset and c) a key - encoded in a form that does match a simple webserver setup serving RIA's default layout out of the box (dirhashmixed). Additionally, it comes with the advantage, that serving archives via a WSGI/CGI is relatively easy, as that script can simply take that path part and look it up in the archive (note, that `export-archive-ora` is intended to run on a clone, hence most likely it produces dirhashmixed-content). Disadvantage, however: When the store layout changes (to dirhashlower as the only currently available alternative for now) the server setup becomes a little complicated as the webserver now needs to translate dirhashmixed to dirhashlower in order to know what to serve.

- Option 2:
Adjust the request URL to the layout the store reports and build the path part of that URL accordingly. Obvious advantage: Switching to dirhashlower or whatever the future may bring is way easier a server setup as it remains a simple serving of what's on the file system. It does, however, make serving archives a little more complicated. That's because archives are likely to have been build from a regular repository and hence containing a dirhashmixed layout. If the store itself uses dirhashlower, then with this second option a store maintainer either has to adjust the archives (which may be a permanent effort, depending on where they are coming from) or the WSGI/CGI has to do the translation in order to perform the lookup in the archive. This would have to be done per request, whereas the translation for non-archived keys in the first option may or may not be something to perform per request - depends.

Finally, anticipating possible future developments, option 1 may be more suitable for more dynamic setups, where a store is a service really that doesn't simply serve an annex object tree from FS.


There may be more implications or not - that's why I wanted to pose the question. The above is, what I'm aware of.

Ultimately, I do lean towards option 2, since it's more straightforward on the server side in what appears to be the "simple and usual case". That may turn out to not be the case, though. 
In case of option 2, we can simply merge your PR and proceed with an RF to make sure `get` and `whereis` don't diverge again. If, however, we stick (or rather go back) to the original approach 1, then your fix needs to be slightly adjusted (`key_path` isn't dynamic in that case) and `get`'s current state reverted to what it was (plus the same RF as follow-up: one ring to bind the two).

**Comment by mih** [comment]
This still needs a fix, after #6091 was shut down. Is this on your list @bpoldrack?

**Comment by bpoldrack** [comment]
@mih 

Yes, if there's no further opinion on option 1 vs. option 2, I'll do that.

**Comment by mih** [comment]
Done with #6091
